File: clients/ios/Sources/SpectreCore/ExprEvaluator.swift
import Foundation

/// 式評価のスコープ。
///
/// `data` は不変、`state` は可変、`item`/`index` は repeat の内側でのみ導入される。
/// `error` は request の onError ハンドラ内でのみ存在する。
public struct EvalScope: Sendable {
    public var data: SpValue
    public var state: SpValue
    public var env: SpValue
    public var locals: [String: SpValue]

    public init(
        data: SpValue = .emptyObject,
        state: SpValue = .emptyObject,
        env: SpValue = .emptyObject,
        locals: [String: SpValue] = [:]
    ) {
        self.data = data
        self.state = state
        self.env = env
        self.locals = locals
    }

    public func withLocals(_ extra: [String: SpValue]) -> EvalScope {
        var copy = self
        copy.locals = locals.merging(extra) { _, new in new }
        return copy
    }

    func lookup(_ name: String) -> SpValue? {
        switch name {
        case "data": return data
        case "state": return state
        case "env": return env
        default: return locals[name]
        }
    }
}

public struct EvalResult: Sendable {
    public let value: SpValue
    public let errors: [ExprError]

    public init(_ value: SpValue, _ errors: [ExprError] = []) {
        self.value = value
        self.errors = errors
    }

    public var hasError: Bool { !errors.isEmpty }
}

/// SpectreExpr の評価器。
///
/// 評価は例外を投げない。エラーは値 `.null` と `ExprError` の記録として現れる
/// (docs/spec/expression.md §5)。壊れた式で画面全体が落ちるより、その部分だけが
/// 空になるほうが害が小さいという判断。
public struct ExprEvaluator {
    private let functions: BuiltinFunctions

    public init(functions: BuiltinFunctions = BuiltinFunctions()) {
        self.functions = functions
    }

    public func evaluate(_ expr: Expr, scope: EvalScope) -> EvalResult {
        var errors: [ExprError] = []
        let value = eval(expr, scope, &errors)
        return EvalResult(value, errors)
    }

    private func eval(_ expr: Expr, _ scope: EvalScope, _ errors: inout [ExprError]) -> SpValue {
        switch expr {
        case .literal(let value):
            return value

        case .identifier(let name):
            return scope.lookup(name) ?? .null

        case .member(let target, let name, _):
            // null や非オブジェクトへのアクセスは null。'.' と '?.' の動作は同じで、
            // '?.' は読み手に意図を伝えるための糖衣構文 (docs/spec/expression.md §3)。
            guard case .object(let entries) = eval(target, scope, &errors) else { return .null }
            return entries[name] ?? .null

        case .index(let target, let indexExpr):
            let base = eval(target, scope, &errors)
            let idx = eval(indexExpr, scope, &errors)
            switch (base, idx) {
            case (.array(let items), .number(let n)):
                let i = Int(n)
                return (i >= 0 && i < items.count) ? items[i] : .null
            case (.object(let entries), .string(let key)):
                return entries[key] ?? .null
            case (.string(let s), .number(let n)):
                let chars = Array(s)
                let i = Int(n)
                return (i >= 0 && i < chars.count) ? .string(String(chars[i])) : .null
            default:
                return .null
            }

        case .call(let name, let args):
            let evaluated = args.map { eval($0, scope, &errors) }
            return functions.invoke(name: name, args: evaluated, scope: scope, errors: &errors)

        case .unary(let op, let operand):
            switch op {
            case "!":
                return .bool(!eval(operand, scope, &errors).isTruthy)
            case "-":
                guard case .number(let v) = eval(operand, scope, &errors) else {
                    return typeError(&errors, "単項 '-' は数値にのみ適用できます")
                }
                return .number(-v)
            default:
                return typeError(&errors, "未知の単項演算子 '\(op)'")
            }

        case .binary(let op, let left, let right):
            return evalBinary(op, left, right, scope, &errors)

        case .ternary(let condition, let ifTrue, let ifFalse):
            return eval(condition, scope, &errors).isTruthy
                ? eval(ifTrue, scope, &errors)
                : eval(ifFalse, scope, &errors)

        case .arrayLiteral(let items):
            return .array(items.map { eval($0, scope, &errors) })

        case .objectLiteral(let pairs):
            var entries: [String: SpValue] = [:]
            for (key, valueExpr) in pairs { entries[key] = eval(valueExpr, scope, &errors) }
            return .object(entries)
        }
    }

    private func evalBinary(
        _ op: String,
        _ leftExpr: Expr,
        _ rightExpr: Expr,
        _ scope: EvalScope,
        _ errors: inout [ExprError]
    ) -> SpValue {
        // && と || は短絡評価する。右辺は必要になるまで評価しない。
        if op == "&&" {
            let left = eval(leftExpr, scope, &errors)
            if !left.isTruthy { return .bool(false) }
            return .bool(eval(rightExpr, scope, &errors).isTruthy)
        }
        if op == "||" {
            let left = eval(leftExpr, scope, &errors)
            if left.isTruthy { return .bool(true) }
            return .bool(eval(rightExpr, scope, &errors).isTruthy)
        }

        let l = eval(leftExpr, scope, &errors)
        let r = eval(rightExpr, scope, &errors)

        switch op {
        case "==": return .bool(l == r)
        case "!=": return .bool(l != r)

        case "+":
            if case .number(let a) = l, case .number(let b) = r { return .number(a + b) }
            // どちらかが文字列なら連結。数値との連結は日常的に必要なため許す。
            let leftText = l.asString != nil, leftNumber = l.asDouble != nil
            let rightText = r.asString != nil, rightNumber = r.asDouble != nil
            if (leftText && (rightText || rightNumber)) || (leftNumber && rightText) {
                return .string(l.stringify() + r.stringify())
            }
            return typeError(&errors, "'+' は数値同士か、文字列と数値/文字列にのみ適用できます")

        case "-", "*", "/", "%":
            guard case .number(let a) = l, case .number(let b) = r else {
                return typeError(&errors, "'\(op)' は数値にのみ適用できます")
            }
            if (op == "/" || op == "%") && b == 0 {
                // 0 除算は例外にせず null。表示が空になるだけで画面は壊れない。
                return typeError(&errors, "0 で除算しました")
            }
            switch op {
            case "-": return .number(a - b)
            case "*": return .number(a * b)
            case "/": return .number(a / b)
            default: return .number(a.truncatingRemainder(dividingBy: b))
            }

        case "<", "<=", ">", ">=":
            return compare(l, r, op, &errors)

        default:
            return typeError(&errors, "未知の二項演算子 '\(op)'")
        }
    }

    private func compare(
        _ l: SpValue,
        _ r: SpValue,
        _ op: String,
        _ errors: inout [ExprError]
    ) -> SpValue {
        let ordering: Int
        switch (l, r) {
        case (.number(let a), .number(let b)):
            ordering = a < b ? -1 : (a > b ? 1 : 0)
        case (.string(let a), .string(let b)):
            ordering = a < b ? -1 : (a > b ? 1 : 0)
        default:
            return typeError(&errors, "'\(op)' は同じ型 (数値同士・文字列同士) にのみ適用できます")
        }
        switch op {
        case "<": return .bool(ordering < 0)
        case "<=": return .bool(ordering <= 0)
        case ">": return .bool(ordering > 0)
        default: return .bool(ordering >= 0)
        }
    }

    private func typeError(_ errors: inout [ExprError], _ message: String) -> SpValue {
        errors.append(ExprError(.type, message))
        return .null
    }
}

File: clients/android/spectre-core/src/main/kotlin/dev/spectre/core/expr/ExprEvaluator.kt
package dev.spectre.core.expr

import dev.spectre.core.SpValue
import dev.spectre.core.isTruthy
import dev.spectre.core.stringify

/**
 * 式評価のスコープ。
 *
 * `data` は不変、`state` は可変、`item`/`index` は repeat の内側でのみ導入される。
 * `error` は request の onError ハンドラ内でのみ存在する。
 */
data class EvalScope(
    val data: SpValue = SpValue.EmptyObj,
    val state: SpValue = SpValue.EmptyObj,
    val env: SpValue = SpValue.EmptyObj,
    val locals: Map<String, SpValue> = emptyMap(),
) {
    fun withLocals(extra: Map<String, SpValue>): EvalScope = copy(locals = locals + extra)

    fun lookup(name: String): SpValue? = when (name) {
        "data" -> data
        "state" -> state
        "env" -> env
        else -> locals[name]
    }
}

data class EvalResult(val value: SpValue, val errors: List<ExprError> = emptyList()) {
    val hasError: Boolean get() = errors.isNotEmpty()
}

/**
 * SpectreExpr の評価器。
 *
 * 評価は例外を投げない。エラーは値 [SpValue.Null] と [ExprError] の記録として現れる
 * (docs/spec/expression.md §5)。壊れた式で画面全体が落ちるより、その部分だけが
 * 空になるほうが害が小さいという判断。
 */
class ExprEvaluator(private val functions: BuiltinFunctions = BuiltinFunctions()) {

    fun evaluate(expr: Expr, scope: EvalScope): EvalResult {
        val errors = ArrayList<ExprError>()
        val value = eval(expr, scope, errors)
        return EvalResult(value, errors)
    }

    private fun eval(expr: Expr, scope: EvalScope, errors: MutableList<ExprError>): SpValue = when (expr) {
        is Expr.Literal -> expr.value

        is Expr.Identifier -> scope.lookup(expr.name) ?: SpValue.Null

        is Expr.Member -> when (val target = eval(expr.target, scope, errors)) {
            is SpValue.Obj -> target.entries[expr.name] ?: SpValue.Null
            // null や非オブジェクトへのアクセスは null。'.' と '?.' の動作は同じで、
            // '?.' は読み手に意図を伝えるための糖衣構文 (docs/spec/expression.md §3)。
            else -> SpValue.Null
        }

        is Expr.Index -> {
            val target = eval(expr.target, scope, errors)
            val index = eval(expr.index, scope, errors)
            when {
                target is SpValue.Arr && index is SpValue.Num -> {
                    val i = index.value.toInt()
                    if (i >= 0 && i < target.items.size) target.items[i] else SpValue.Null
                }
                target is SpValue.Obj && index is SpValue.Str -> target.entries[index.value] ?: SpValue.Null
                target is SpValue.Str && index is SpValue.Num -> {
                    val i = index.value.toInt()
                    if (i >= 0 && i < target.value.length) SpValue.Str(target.value[i].toString()) else SpValue.Null
                }
                else -> SpValue.Null
            }
        }

        is Expr.Call -> {
            val args = expr.args.map { eval(it, scope, errors) }
            functions.invoke(expr.name, args, scope, errors)
        }

        is Expr.Unary -> when (expr.op) {
            "!" -> SpValue.Bool(!eval(expr.operand, scope, errors).isTruthy)
            "-" -> when (val v = eval(expr.operand, scope, errors)) {
                is SpValue.Num -> SpValue.Num(-v.value)
                else -> typeError(errors, "単項 '-' は数値にのみ適用できます")
            }
            else -> typeError(errors, "未知の単項演算子 '${expr.op}'")
        }

        is Expr.Binary -> evalBinary(expr, scope, errors)

        is Expr.Ternary ->
            if (eval(expr.condition, scope, errors).isTruthy) eval(expr.ifTrue, scope, errors)
            else eval(expr.ifFalse, scope, errors)

        is Expr.ArrayLit -> SpValue.Arr(expr.items.map { eval(it, scope, errors) })

        is Expr.ObjectLit -> SpValue.Obj(
            expr.entries.associate { (k, v) -> k to eval(v, scope, errors) }
        )
    }

    private fun evalBinary(expr: Expr.Binary, scope: EvalScope, errors: MutableList<ExprError>): SpValue {
        // && と || は短絡評価する。右辺は必要になるまで評価しない。
        if (expr.op == "&&") {
            val left = eval(expr.left, scope, errors)
            if (!left.isTruthy) return SpValue.Bool(false)
            return SpValue.Bool(eval(expr.right, scope, errors).isTruthy)
        }
        if (expr.op == "||") {
            val left = eval(expr.left, scope, errors)
            if (left.isTruthy) return SpValue.Bool(true)
            return SpValue.Bool(eval(expr.right, scope, errors).isTruthy)
        }

        val l = eval(expr.left, scope, errors)
        val r = eval(expr.right, scope, errors)

        return when (expr.op) {
            "==" -> SpValue.Bool(deepEquals(l, r))
            "!=" -> SpValue.Bool(!deepEquals(l, r))

            "+" -> when {
                l is SpValue.Num && r is SpValue.Num -> SpValue.Num(l.value + r.value)
                // どちらかが文字列なら連結。数値との連結は日常的に必要なため許す。
                l is SpValue.Str && (r is SpValue.Str || r is SpValue.Num) ->
                    SpValue.Str(l.value + r.stringify())
                r is SpValue.Str && l is SpValue.Num -> SpValue.Str(l.stringify() + r.value)
                else -> typeError(errors, "'+' は数値同士か、文字列と数値/文字列にのみ適用できます")
            }

            "-", "*", "/", "%" -> {
                if (l !is SpValue.Num || r !is SpValue.Num) {
                    typeError(errors, "'${expr.op}' は数値にのみ適用できます")
                } else if ((expr.op == "/" || expr.op == "%") && r.value == 0.0) {
                    // 0 除算は例外にせず null。表示が空になるだけで画面は壊れない。
                    typeError(errors, "0 で除算しました")
                } else {
                    SpValue.Num(
                        when (expr.op) {
                            "-" -> l.value - r.value
                            "*" -> l.value * r.value
                            "/" -> l.value / r.value
                            else -> l.value % r.value
                        }
                    )
                }
            }

            "<", "<=", ">", ">=" -> compare(l, r, expr.op, errors)

            else -> typeError(errors, "未知の二項演算子 '${expr.op}'")
        }
    }

    private fun compare(
        l: SpValue,
        r: SpValue,
        op: String,
        errors: MutableList<ExprError>,
    ): SpValue {
        val cmp: Int = when {
            l is SpValue.Num && r is SpValue.Num -> l.value.compareTo(r.value)
            l is SpValue.Str && r is SpValue.Str -> l.value.compareTo(r.value)
            else -> return typeError(errors, "'$op' は同じ型 (数値同士・文字列同士) にのみ適用できます")
        }
        return SpValue.Bool(
            when (op) {
                "<" -> cmp < 0
                "<=" -> cmp <= 0
                ">" -> cmp > 0
                else -> cmp >= 0
            }
        )
    }

    private fun typeError(errors: MutableList<ExprError>, message: String): SpValue {
        errors.add(ExprError(ExprErrorCode.E_TYPE, message))
        return SpValue.Null
    }
}

/** 型変換をしない厳密比較。配列/オブジェクトは構造的に比較する。 */
fun deepEquals(a: SpValue, b: SpValue): Boolean = when {
    a is SpValue.Null && b is SpValue.Null -> true
    a is SpValue.Bool && b is SpValue.Bool -> a.value == b.value
    a is SpValue.Num && b is SpValue.Num -> a.value == b.value
    a is SpValue.Str && b is SpValue.Str -> a.value == b.value
    a is SpValue.Arr && b is SpValue.Arr ->
        a.items.size == b.items.size && a.items.indices.all { deepEquals(a.items[it], b.items[it]) }
    a is SpValue.Obj && b is SpValue.Obj ->
        a.entries.keys == b.entries.keys &&
            a.entries.all { (k, v) -> deepEquals(v, b.entries.getValue(k)) }
    else -> false
}
